lua/nvim-treesitter/fold.lua

```diff
@@ -1,28 +1,60 @@
+local api = vim.api
+local utils = require'nvim-treesitter.ts_utils'
+local query = require'nvim-treesitter.query'
 local parsers = require'nvim-treesitter.parsers'
 
 local M = {}
 
-function M.get_fold_indic(lnum)
-  if not parsers.has_parser() or not lnum then return '0' end
+local folds_levels = utils.memoize_by_buf_tick(function(bufnr)
+  print("hey")
+  local levels_tmp = {}
 
-  local function smallest_multiline_containing(node, level)
-    for index = 0,(node:named_child_count() -1) do
-      local child = node:named_child(index)
-      local start, _, stop, _ = child:range()
+  for _, node in ipairs(query.get_capture_matches(bufnr, "fold", "fold")) do
+    print("yo")
+    local start, _, stop, _ = node:range()
 
-      if start ~= stop and start <= (lnum -1) and stop >= (lnum -1) then
-        return smallest_multiline_containing(child, level + 1)
-      end
+    -- This can be folded
+    -- Fold only multiline nodes that are not exactly the same as prevsiously met folds
+    if start ~= stop and not (levels_tmp[start] and levels_tmp[stop]) then
+      levels_tmp[start] = (levels_tmp[start] or 0) + 1
+      levels_tmp[stop] = (levels_tmp[stop] or 0) - 1
     end
+  end
+
+  local levels = {}
+  local current_level = 0
 
-    return node, level
+  for lnum=0,api.nvim_buf_line_count(bufnr) do
+    current_level = current_level + levels_tmp[lnum]
+    levels[lnum + 1] = current_level
   end
 
-  local parser = parsers.get_parser()
+  print("levels", vim.inspect(levels))
+  return levels
+end)
+
+function M.get_fold_indic(lnum)
+  if not parsers.has_parser() or not lnum then return '0' end
 
-  local _, level = smallest_multiline_containing(parser:parse():root(), 0)
+  local buf = api.nvim_get_current_buf()
+
+  print(vim.inspect(folds_levels))
+
+  local levels = folds_levels(buf) or {}
+
+  return tostring(levels[lnum] or 0)
+end
+
+function M.attach(bufnr)
+  local win = vim.fn.win_findbuf(bufnr)[1]
```

Comment:
A buffer can have multiple windows.

Reply:
Will the foldmethod be inherited? When the window switches to a different buffer it will still have the folding marker. This why I gave up my try to make folding a module. But thanks that you are tackling the problem.

Reply:
Oh, that's an interesting issue

Reply:
Maybe we can have the module enabled globally by default, and let users fine tune it using `nvim_treesitter#foldexr()` otherwise